## Project Title: Aspect-Based Sentiment Analysis for Product Review Insights

### Introduction

In modern e-commerce platforms such as Amazon, Flipkart, and Myntra, users purchase a wide range of products and share their experiences through reviews. These reviews contain valuable information about customer satisfaction and product quality.

However, manually analyzing a large volume of reviews to understand overall product performance is time-consuming and inefficient. Traditional sentiment analysis techniques classify reviews into categories such as positive, negative, or neutral, but they fail to provide detailed insights about specific product features.

To overcome this limitation, this project focuses on **Aspect-Based Sentiment Analysis (ABSA)**, which enables fine-grained analysis by identifying individual product aspects and determining the sentiment associated with each of them.

---

### Objective

The primary objective of this project is to develop a Natural Language Processing (NLP) model that can:

* Automatically identify key aspects of a product from customer reviews
* Determine the sentiment (positive, negative, or neutral) associated with each aspect
* Provide structured insights into product strengths and weaknesses based on user feedback

This approach helps in generating more meaningful and actionable insights compared to traditional sentiment analysis methods.


In [1]:
#Importing Libraries
import nltk
import pandas as pd

### Step 1: Data Collection

For this project, we collected a real-world dataset of Amazon product reviews from Google Dataset Search.

The dataset consists of **4,916 rows and 12 columns**, containing detailed information such as review text, ratings, and product identifiers.

From this dataset, we filtered and selected reviews corresponding to a specific product using its ASIN (Amazon Standard Identification Number). The selected product is a **SanDisk microSD Memory Card**, which is a storage device commonly used in smartphones, cameras, and other electronic devices.

Focusing on a single product helps ensure consistency in the type of aspects (such as speed, storage capacity, and performance) mentioned in the reviews, making the aspect-based sentiment analysis more accurate and meaningful.


In [3]:
#Step-2: Data Loading

pd_df=pd.read_csv("amazon_review.csv")
pd_df

,reviewerID,asin,reviewerName,helpful,reviewText,overall,summary,unixReviewTime,reviewTime,day_diff,helpful_yes,total_vote
0,A3SBTW3WS4IQSN,B007WTAJTO,NaN,"[0, 0]",No issues.,4.0,Four Stars,1406073600,2014-07-23,138,0,0
1,A18K1ODH1I2MVB,B007WTAJTO,0mie,"[0, 0]","Purchased this for my device, it worked as adv...",5.0,MOAR SPACE!!!,1382659200,2013-10-25,409,0,0
2,A2FII3I2MBMUIA,B007WTAJTO,1K3,"[0, 0]",it works as expected. I should have sprung for...,4.0,nothing to really say....,1356220800,2012-12-23,715,0,0
3,A3H99DFEG68SR,B007WTAJTO,1m2,"[0, 0]",This think has worked out great.Had a diff. br...,5.0,Great buy at this price!!! *** UPDATE,1384992000,2013-11-21,382,0,0
4,A375ZM4U047O79,B007WTAJTO,2&amp;1/2Men,"[0, 0]","Bought it with Retail Packaging, arrived legit...",5.0,best deal around,1373673600,2013-07-13,513,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...
4910,A2LBMKXRM5H2W9,B007WTAJTO,"ZM ""J""","[0, 0]",I bought this Sandisk 16GB Class 10 to use wit...,1.0,Do not waste your money.,1374537600,2013-07-23,503,0,0
4911,ALGDLRUI1ZPCS,B007WTAJTO,Zo,"[0, 0]",Used this for extending the capabilities of my...,5.0,Great item!,1377129600,2013-08-22,473,0,0
4912,A2MR1NI0ENW2AD,B007WTAJTO,Z S Liske,"[0, 0]",Great card that is very fast and reliable. It ...,5.0,Fast and reliable memory card,1396224000,2014-03-31,252,0,0
4913,A37E6P3DSO9QJD,B007WTAJTO,Z Taylor,"[0, 0]",Good amount of space for the stuff I want to d...,5.0,Great little card,1379289600,2013-09-16,448,0,0


In [4]:
#Step-3: Data Inspection

#1)head:It returns us the first 5 rows of the dataset.
pd_df.head()

#2)tail: It returns us the last 5 rows of the dataset.
pd_df.tail()

#3)info():It tells us what columns are present in the dataset and their datatype and nn-null count.
pd_df.info()

#4)describe():It gives us the statistcis of numerical columns such as mean,std,median etc..
print(pd_df.describe())
print(pd_df.describe(include=["object"]))

#5)columns:It gives the column names present in the dataset
pd_df.columns

#6)shape:It gives us the matrix of no of rows and columns present in the dataset.
pd_df.shape

#7)isnull().sum():It gives the total number of missing values present in a every column individually.
pd_df.isnull().sum()

#8)duplicated().sum():It tells the number of duplicate rows present over the dataset.
pd_df.duplicated().sum()

#9)nunique():It gives us the no of unique values present in the dataset.
pd_df.nunique()

#10)value_counts(): It gives the distribution of the dataset and from this we check whether the dataset is balanced or imbalnced.
pd_df["overall"].value_counts()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4915 entries, 0 to 4914
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   reviewerID      4915 non-null   object 
 1   asin            4915 non-null   object 
 2   reviewerName    4914 non-null   object 
 3   helpful         4915 non-null   object 
 4   reviewText      4914 non-null   object 
 5   overall         4915 non-null   float64
 6   summary         4915 non-null   object 
 7   unixReviewTime  4915 non-null   int64  
 8   reviewTime      4915 non-null   object 
 9   day_diff        4915 non-null   int64  
 10  helpful_yes     4915 non-null   int64  
 11  total_vote      4915 non-null   int64  
dtypes: float64(1), int64(4), object(7)
memory usage: 460.9+ KB
           overall  unixReviewTime     day_diff  helpful_yes   total_vote
count  4915.000000    4.915000e+03  4915.000000  4915.000000  4915.000000
mean      4.587589    1.379465e+09   437.367040    

,count
overall,
5.0,3922
4.0,527
1.0,244
3.0,142
2.0,80


Insights from Data Inspection:

1. From head we have checked that the columns names are present rather than any row specification and we have checked even that the data is loaded successfully.
2. From tail we have got a clarity that the entire dataset is loaded.

3. From info we get to know what type of columns[i.e;what type of values a specific column is storing] and then we have observed that there is  one missing value in both columns reviewername and reviewer text.

4. from columns we get to know what are the columns present in the dataset.

5. From shape we get to know the shape of the dataset[number of rows and columns present in the dataset] which helps in decision making of splitting the dataset.

6. describe: we get to see the statisitcs from of the numerical columns and categorical columns.
From this we get to know that reviewtime is wrongly stored as object but it has to be datatime datatype.
7. There are zero duplicate rows and 1 missing value in reviewername and reviewer text.

8. The nunique() function shows the number of unique values in each column. It helps in understanding the diversity of data. For example, the ASIN column contains only one unique value, confirming that all reviews belong to a single product.

9. From value_counts we have observed that the dataset is imbalnced where the distribution is not equal there are 3922 rows of 5 rating which is almost 79% of the dataset it is completely biased. this can lead to overfitting to aviod this:
1)We use  balanced in the model to priortize everything equally
2)And in evalution metrics we do not consider accuracy much rather we depend on f1 score bcz here the cost of fp and fn are almost equal so yes we go up with f1-score for the balnce between them